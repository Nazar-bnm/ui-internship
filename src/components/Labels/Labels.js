import React from 'react';

import './Labels.scss';

const Labels = (props) => {
  const { blogItems, getLabel } = props;
  const listOfLabels = blogItems?.map((el) => el.labels)?.flat()?.sort();
  const countedLabels = {};

  listOfLabels && listOfLabels.forEach((el) => {
    if (!countedLabels[el]) {
      countedLabels[el] = 1;
    } else {
      countedLabels[el] += 1;
    }
  });

  const maxLabelCount = Math.max(...Object.values(countedLabels));
  const elementToRender = (elem, fontSize) => (<span className="label__item" style={{ fontSize: `${fontSize}rem` }}>{elem}</span>);
  const divideInto = (num) => Math.round(maxLabelCount / num);
  const formLables = (list) => Object.keys(list).map((el) => {
    // There are 4 different types of styles therefore further I use numbers to dived on 2, 3, 4
    const fontSizeforLargeEl = 1.6;
    const fontSizeForMediumEl = 1.3;
    const fontSizeForSmallEl = 1;
    const fontSizeForExtraSmallEl = 0.7;

    if (list[el] >= divideInto(2)) {
      return elementToRender(el, fontSizeforLargeEl);
    } if (list[el] >= divideInto(3) && list[el] <= divideInto(2)) {
      return elementToRender(el, fontSizeForMediumEl);
    } if (list[el] >= Math.round(maxLabelCount / 4) && list[el] <= divideInto(3)) {
      return elementToRender(el, fontSizeForSmallEl);
    }

    return elementToRender(el, fontSizeForExtraSmallEl);
  });

  const handleClick = ({ target }) => {
    const { innerText } = target;
    getLabel(innerText.toLowerCase());
  };

  return (
    <>
      <div className="label__sectionTitle">Labels</div>
      <div className="label" onClick={handleClick}>{formLables(countedLabels)}</div>
    </>
  );
};

export default Labels;
