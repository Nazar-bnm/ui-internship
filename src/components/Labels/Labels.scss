@import '../../scss/global_styles';

.label {
  display: flex;
  flex-wrap: wrap;
  cursor: pointer;

  &__item{
    padding: 5px;
    text-transform: uppercase;
    align-self: flex-end;
  }

  &__sectionTitle {
    text-transform: uppercase;
    margin-bottom: 15px;
  }
}