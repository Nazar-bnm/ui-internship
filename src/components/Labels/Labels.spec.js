import React from 'react';
import { shallow } from 'enzyme';

import Labels from './Labels';

describe('<Labels />', () => {
  let wrapper;
  let props;
  const handleClick = jest.fn();

  beforeEach(() => {
    props = {
      getLabel: jest.fn()
    }
    wrapper = shallow(
      <Labels {...props}/>
    );
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should invoke function handleClick when event on click occurs', () => {
    const event = { target: { innerText: 'TEST' } };
    wrapper
      .find('.label')
      .simulate('click', event);
    // expect(handleClick.mock.calls.length).toBe(1);
    expect(props.getLabel).toHaveBeenCalledWith('test');
  });
});
