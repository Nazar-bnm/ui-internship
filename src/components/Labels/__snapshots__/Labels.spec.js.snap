// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Labels /> should match snapshot 1`] = `
<Fragment>
  <div
    className="label__sectionTitle"
  >
    Labels
  </div>
  <div
    className="label"
    onClick={[Function]}
  />
</Fragment>
`;
